core/integration/kafka/plugin.go

```diff
@@ -0,0 +1,311 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In applying this license CERN does not waive the privileges and
+ * immunities granted to it by virtue of its status as an
+ * Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+//go:generate protoc --go_out=. protos/kafka.proto
+
+package kafka
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/AliceO2Group/Control/common/logger"
+	"github.com/AliceO2Group/Control/common/utils/uid"
+	"github.com/AliceO2Group/Control/core/integration"
+	kafkapb "github.com/AliceO2Group/Control/core/integration/kafka/protos"
+	"github.com/AliceO2Group/Control/core/workflow/callable"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/golang/protobuf/proto"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var log = logger.New(logrus.StandardLogger(), "kafkaplugin")
+
+type Plugin struct {
+	kafkaBroker   string
+	kafkaProducer *kafka.Producer
+	envsInRunning map[string]*kafkapb.EnvInfo // env id is the key
+}
+
+func NewPlugin(endpoint string) integration.Plugin {
+	_, err := url.Parse(endpoint)
+	if err != nil {
+		log.WithField("endpoint", endpoint).
+			WithError(err).
+			Error("bad service endpoint")
+		return nil
+	}
+
+	return &Plugin{
+		kafkaBroker:   endpoint,
+		kafkaProducer: nil,
+	}
+}
+
+func (p *Plugin) GetName() string {
+	return "kafka"
+}
+
+func (p *Plugin) GetPrettyName() string {
+	return "Kafka FSM transition information service"
+}
+
+func (p *Plugin) GetEndpoint() string {
+	return viper.GetString("kafkaBroker")
+}
+
+func (p *Plugin) GetConnectionState() string {
+	if p == nil || p.kafkaProducer == nil {
+		return "UNKNOWN"
+	} else {
+		return "READY"
+	}
+}
+
+func (p *Plugin) GetData(environmentIds []uid.ID) string {
+	return ""
+}
+
+func (p *Plugin) FSMTransitionTopic(state string) string {
+	return "aliecs.env_state." + state
+}
+
+func (p *Plugin) ActiveRunsListTopic() string {
+	return "aliecs.env_list.RUNNING"
+}
+
+func (p *Plugin) Init(instanceId string) error {
+	var err error
+	p.kafkaProducer, err = kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": p.kafkaBroker})
+	if err != nil {
+		return fmt.Errorf("failed to initialize a kafka producer with broker '%s'. Details: %s", p.kafkaBroker, err.Error())
+	}
+	p.envsInRunning = make(map[string]*kafkapb.EnvInfo)
+	log.Info("Successfully created a kafka producer with broker '" + p.kafkaBroker + "'")
+
+	// Prepare and send active run list (expected to be empty during init)
+	timestamp := uint64(time.Now().UnixMilli())
+	activeRunsList := &kafkapb.ActiveRunsList{
+		ActiveRuns: p.GetRunningEnvList(),
+		Timestamp:  timestamp,
+	}
+	arlData, err := proto.Marshal(activeRunsList)
+	if err != nil {
+		log.Error("Could not marshall an active runs list: ", err)
+	}
+	p.ProduceMessage(arlData, p.ActiveRunsListTopic(), "")
```

Comment:
I suspect some of these member functions could be made private (uncapitalized initial)